src/spi2_pool.rs

```diff
@@ -0,0 +1,211 @@
+use crate::gpio::{AnyOutputPin, InputPin, Output, OutputPin, PinDriver};
+use crate::peripheral::{Peripheral, PeripheralRef};
+use crate::spi2::{config, SpiBusMasterDriver, SpiMasterDriver};
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use core::ptr;
+
+use esp_idf_sys::*;
+
+// TODO -> custom error that wraps around esperror
+// cs default high low setting -
+// needed to makte SpiBusMasterDriver fields pub because different module
+// needed to make SpiBusMasterDriver flush fn pub
+
+// performance measurment on esp32c3 -> 
+// debug build: 50 uS deley between cs low -> first clk out
+// release build : 15 uS deley between cs low -> first cll out
+
+// debug build: 4uS delay between last clk out -> cs pin high
+// release build: 3uS delay between last clk out -> cs pin high
+
+#[cfg(not(esp32c3))]
+const ESP_MAX_SPI_DEVICES: usize = 3;
+#[cfg(esp32c3)]
+const ESP_MAX_SPI_DEVICES: usize = 6;
+
+pub struct SpiConfigPool<'d, T> {
+    shared_handles: HashMap<u32, spi_device_handle_t>,
+    master: T,
+    _p: PhantomData<&'d ()>,
+}
```

Comment:
I see you're using the raw `spi_device_handle_t`, were you not able to do this?
```Rust
pub struct SpiConfigPool<'d, T> {
    shared_handles: HashMap<u32, EspSpiDevice<'d, T>>,
    master: T,
    _p: PhantomData<&'d ()>,
}
```

Reply:
> I see you're using the raw `spi_device_handle_t`, were you not able to do this?
> 
> ```rust
> pub struct SpiConfigPool<'d, T> {
>     shared_handles: HashMap<u32, EspSpiDevice<'d, T>>,
>     master: T,
>     _p: PhantomData<&'d ()>,
> }
> ```
yeah that was the initial plan. i switched it back here to the raw handle because of some design decision 
- because of the cs_glitches i got when i frequently removed an device handle with spi_bus_remove_device() i currently opted out of the idea of switching out the spi_device_handle_t actively ( this is a deeper dive first):
- for regular usage one can now use SpiMasterDriver with EspSpiDevice witch uses hardware cs
- redefining the purpose of the SpiConfigPool: The SpiConfigPool gets you up to 3/6 independent "transmission configs" ( an spi_device_handle_t). 
    - because each "transmission configs" is now shared it should not allow to have an cs pin set to anything -> an EspSpiDevice allows to set an cs_pin, 
    - its probably less resource intense to just store just the handle 
    - the transaction from EspSpiDevice and the SpiPoolDevice are different. The First must call SpiBusMasterDriver with hardware_cs true / the later with false. SpiPoolDevice needs to manage the cs pin here now actively -> want the high/low calls close to the actual transaction. but yes yours look much nicer without so much duplication :smiley_cat: 
 - because of that i didn't think about actually using an EspSpiDevice here

Reply:
Ah okay I see. So for the new idea, what's the different between this and the simple hardware CS option? Does this give you unlimited devices but subsets of them share a `spi_device_handle_t`?

Reply:
> Ah okay I see. So for the new idea, what's the different between this and the simple hardware CS option? Does this give you unlimited devices but subsets of them share a `spi_device_handle_t`?
yes that's the basic idea. but the nice thing is you can use it in two ways here. the first is unlimited devices. but you can also use it for the reverce. if you have one device that needed different configs you can now just change configs without switching the actual config / handle of the device. an config switch is normaly also an handle switch witch lets the cs_pin in an glitchy state currently. this does prevent it. want a different config to your device? just change the config_id in your SpiPoolDevice to one of the 3/6